glommio/src/sync/semaphore.rs

```diff
@@ -7,49 +7,115 @@ use std::cell::RefCell;
 }
 
-#[derive(Debug)]
-struct WaiterState {
-    units: u64,
-    waker: Option<Waker>,
+struct WaiterPointerOps;
+
+unsafe impl PointerOps for WaiterPointerOps {
+    type Value = WaiterNode;
+    type Pointer = NonNull<WaiterNode>;
+
+    unsafe fn from_raw(&self, value: *const Self::Value) -> Self::Pointer {
+        NonNull::new_unchecked(value as *mut Self::Value)
+    }
+
+    fn into_raw(&self, ptr: Self::Pointer) -> *const Self::Value {
+        ptr.as_ptr() as *const Self::Value
+    }
 }
 
-intrusive_adapter!(WaiterAdapter = Rc<WaiterNode> : WaiterNode {link : LinkedListLink});
+struct WaiterAdapter {
+    pointers_ops: WaiterPointerOps,
+    link_ops: LinkOps,
+}
+
+impl WaiterAdapter {
+    fn new() -> Self {
+        WaiterAdapter {
+            pointers_ops: WaiterPointerOps,
+            link_ops: LinkOps,
+        }
+    }
+}
+
+unsafe impl Adapter for WaiterAdapter {
+    type LinkOps = LinkOps;
+    type PointerOps = WaiterPointerOps;
+
+    unsafe fn get_value(
+        &self,
+        link: <Self::LinkOps as intrusive_collections::LinkOps>::LinkPtr,
+    ) -> *const <Self::PointerOps as PointerOps>::Value {
+        container_of!(link.as_ptr(), WaiterNode, link)
+    }
 
-impl Waiter {
-    fn new(units: u64, sem_state: Rc<RefCell<SemaphoreState>>) -> Waiter {
+    unsafe fn get_link(
+        &self,
+        value: *const <Self::PointerOps as PointerOps>::Value,
+    ) -> <Self::LinkOps as intrusive_collections::LinkOps>::LinkPtr {
+        let ptr = (value as *const u8).add(offset_of!(WaiterNode, link));
+        core::ptr::NonNull::new_unchecked(ptr as *mut _)
+    }
+
+    fn link_ops(&self) -> &Self::LinkOps {
+        &self.link_ops
+    }
+
+    fn link_ops_mut(&mut self) -> &mut Self::LinkOps {
+        &mut self.link_ops
+    }
+
+    fn pointer_ops(&self) -> &Self::PointerOps {
+        &self.pointers_ops
+    }
+}
+
+impl<'a> Waiter<'a> {
+    fn new(units: u64, semaphore: &'a Semaphore) -> Waiter<'a> {
         Waiter {
-            sem_state,
-            node: Rc::new(WaiterNode {
-                link: LinkedListLink::default(),
-                state: RefCell::new(WaiterState { units, waker: None }),
-            }),
+            node: WaiterNode {
+                link: LinkedListLink::new(),
+                units,
+                waker: RefCell::new(None),
+                _p: PhantomPinned,
+            },
+            semaphore,
         }
     }
 
-    fn remove_from_waiting_queue(&self, sem_state: &mut SemaphoreState) {
-        if self.node.link.is_linked() {
+    fn remove_from_waiting_queue(
+        waiter_node: Pin<&mut WaiterNode>,
+        sem_state: &mut SemaphoreState,
+    ) {
+        if waiter_node.link.is_linked() {
             let mut cursor = unsafe {
                 sem_state
                     .waiters_list
-                    .cursor_mut_from_ptr(self.node.as_ref())
+                    .cursor_mut_from_ptr(Pin::into_inner_unchecked(waiter_node) as *const _)
```

Comment:
I am a bit concerned with this.
Right now usage seems safe, but I worry about future changes.

The reason this seems safe to me is that when we use node, it is only at the time we are ready to destruct it.
But to do that, we do dereference it, and it isn't hard to imagine someone makinga mistake and moving out of it.

Could we implement our our `Deref` for those that just panic, and use a helper function to convert node to its internal components that we are interested at?

Something like `fn consume_node(node: *const WaiterNode) -> (Option<Waker>, u64)`?

So we would never deref -> always safe on the unpin, and still extract the stuff we want ourselves when time comes to use

Reply:
Hi @glommer , we use a pointer to the node here, only to remove from the list we do not access waker. But I suppose you mean that when we access the internal object pointed by the pointer, we should deref it. To mitigate which risks I am a bit confused. There are two risks there to have invalid pointers:
1. Node is moved to other location of memory, so the pointer becomes invalid. That is mitigated because of the object itself  !Unpin and wrapped into the PIn object. Code is unsafe there but intrusive_collections do not move objects. So it is logically safe.
2. Future is dropped from the stack or heap, in such case pointer also becomes invalid. To prevent that we implement Drop trait which removes a node from the list, ensuring that invalid pointers are not stored into the collection.

Could you elaborate what risks do you think will be mitigated by the introduction of this method, because I think only about null checks but they can not happen because we create Pin-s from references?

Reply:
Also this method `cursor_mut_from_ptr` accepts pointer not ref, it only looked like it accepts ref because of implicit conversion.

Reply:
the `Pin` implementation, especially for things that don't implement `Unpin`, require objects to be not movable.
Once you do this: 

```
let node = unsafe { &*node.as_ptr() }; 
```

You now turned your pinned object into a normal object from the point of view of the compiler. It could be moved and `Pin` would break.

Your implementation looks fine to me, because we only do that when we are exhausting the list.
My concern is more with the future proofing of this.

What I suggested would be a way to make sure that we can never dereference that pointer directly: if we can't dereference it, we can't break the `Pin` guarantees. We still need to manipulate it once it is going away and that is what the helper function would be for.

Reply:
Also when we access object, we either access it through the collection, in such case, it performs conversion itself from Pointer (NonNull in our case, that is why we implement PtrOps, to tell it how to convert Pointer to Node) or when we remove it from the node. In such a case, NonNull is returned. In both cases inside PtrOps or inside of the semaphore itself null check is performed by usage of NonNull contract.